File: backend/src/runtime/production_cutover.py
"""Phase 10 — configurable production cutover / alert thresholds.

Thresholds are env-driven so ops can agree SLOs without code changes.
Defaults are conservative starting points, not hard product SLAs.
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Literal


def _env_float(name: str, default: float) -> float:
    raw = os.getenv(name)
    if raw is None or not str(raw).strip():
        return default
    try:
        return float(raw)
    except ValueError:
        return default


def _env_int(name: str, default: int) -> int:
    raw = os.getenv(name)
    if raw is None or not str(raw).strip():
        return default
    try:
        return int(raw)
    except ValueError:
        return default


@dataclass(frozen=True, slots=True)
class ProductionCutoverThresholds:
    min_success_rate: float
    max_error_rate: float
    max_duplicate_rate: float
    max_stale_rate: float
    min_recovery_success_rate: float
    queue_depth_warning: int
    queue_depth_critical: int
    oldest_pending_age_warning_sec: int
    oldest_pending_age_critical_sec: int
    max_p95_job_duration_ms: int


def load_production_cutover_thresholds() -> ProductionCutoverThresholds:
    return ProductionCutoverThresholds(
        min_success_rate=_env_float("PRODUCTION_CUTOVER_MIN_SUCCESS_RATE", 0.99),
        max_error_rate=_env_float("PRODUCTION_CUTOVER_MAX_ERROR_RATE", 0.01),
        max_duplicate_rate=_env_float("PRODUCTION_CUTOVER_MAX_DUPLICATE_RATE", 0.0),
        max_stale_rate=_env_float("PRODUCTION_CUTOVER_MAX_STALE_RATE", 0.02),
        min_recovery_success_rate=_env_float(
            "PRODUCTION_CUTOVER_MIN_RECOVERY_SUCCESS_RATE", 0.99
        ),
        queue_depth_warning=_env_int("PRODUCTION_ALERT_QUEUE_DEPTH_WARNING", 100),
        queue_depth_critical=_env_int("PRODUCTION_ALERT_QUEUE_DEPTH_CRITICAL", 500),
        oldest_pending_age_warning_sec=_env_int(
            "PRODUCTION_ALERT_OLDEST_PENDING_WARNING_SEC", 900
        ),
        oldest_pending_age_critical_sec=_env_int(
            "PRODUCTION_ALERT_OLDEST_PENDING_CRITICAL_SEC", 3600
        ),
        max_p95_job_duration_ms=_env_int("PRODUCTION_CUTOVER_MAX_P95_JOB_MS", 0),
    )


RolloutDecision = Literal["continue", "pause_warning", "pause_critical"]


@dataclass(frozen=True, slots=True)
class RolloutPauseSignals:
    success_rate: float
    error_rate: float
    duplicate_rate: float
    stale_rate: float
    recovery_success_rate: float
    queue_depth: int | None = None
    oldest_pending_age_sec: int | None = None
    p95_job_duration_ms: int | None = None


@dataclass(frozen=True, slots=True)
class RolloutPauseResult:
    decision: RolloutDecision
    reasons: tuple[str, ...]


def evaluate_rollout_pause(
    thresholds: ProductionCutoverThresholds,
    signals: RolloutPauseSignals,
) -> RolloutPauseResult:
    reasons: list[str] = []
    decision: RolloutDecision = "continue"

    def bump(level: RolloutDecision, reason: str) -> None:
        nonlocal decision
        reasons.append(reason)
        if level == "pause_critical":
            decision = "pause_critical"
        elif decision == "continue":
            decision = level

    if signals.success_rate < thresholds.min_success_rate:
        bump(
            "pause_critical",
            f"success_rate {signals.success_rate} < {thresholds.min_success_rate}",
        )
    if signals.error_rate > thresholds.max_error_rate:
        bump(
            "pause_critical",
            f"error_rate {signals.error_rate} > {thresholds.max_error_rate}",
        )
    if signals.duplicate_rate > thresholds.max_duplicate_rate:
        bump(
            "pause_critical",
            f"duplicate_rate {signals.duplicate_rate} > {thresholds.max_duplicate_rate}",
        )
    if signals.stale_rate > thresholds.max_stale_rate:
        bump(
            "pause_warning",
            f"stale_rate {signals.stale_rate} > {thresholds.max_stale_rate}",
        )
    if signals.recovery_success_rate < thresholds.min_recovery_success_rate:
        bump(
            "pause_critical",
            f"recovery_success_rate {signals.recovery_success_rate} < "
            f"{thresholds.min_recovery_success_rate}",
        )
    if signals.queue_depth is not None:
        if signals.queue_depth >= thresholds.queue_depth_critical:
            bump(
                "pause_critical",
                f"queue_depth {signals.queue_depth} >= {thresholds.queue_depth_critical}",
            )
        elif signals.queue_depth >= thresholds.queue_depth_warning:
            bump(
                "pause_warning",
                f"queue_depth {signals.queue_depth} >= {thresholds.queue_depth_warning}",
            )
    if signals.oldest_pending_age_sec is not None:
        if signals.oldest_pending_age_sec >= thresholds.oldest_pending_age_critical_sec:
            bump(
                "pause_critical",
                f"oldest_pending_age_sec {signals.oldest_pending_age_sec} >= "
                f"{thresholds.oldest_pending_age_critical_sec}",
            )
        elif signals.oldest_pending_age_sec >= thresholds.oldest_pending_age_warning_sec:
            bump(
                "pause_warning",
                f"oldest_pending_age_sec {signals.oldest_pending_age_sec} >= "
                f"{thresholds.oldest_pending_age_warning_sec}",
            )
    if (
        thresholds.max_p95_job_duration_ms > 0
        and signals.p95_job_duration_ms is not None
        and signals.p95_job_duration_ms > thresholds.max_p95_job_duration_ms
    ):
        bump(
            "pause_warning",
            f"p95_job_duration_ms {signals.p95_job_duration_ms} > "
            f"{thresholds.max_p95_job_duration_ms}",
        )

    return RolloutPauseResult(decision=decision, reasons=tuple(reasons))

File: mobile/src/core/productionCutover.ts
/**
 * Phase 10 — configurable cutover / pause thresholds (no hardcoded ops SLOs).
 * Values come from Expo `extra.cutover` or process.env in tests/CI.
 */

export interface ProductionCutoverThresholds {
  /** Minimum success rate (0–1) to continue rollout. */
  readonly minSuccessRate: number;
  /** Maximum error rate (0–1) before auto-pause. */
  readonly maxErrorRate: number;
  /** Maximum duplicate job rate (0–1); prefer 0. */
  readonly maxDuplicateRate: number;
  /** Maximum stale conflict rate (0–1). */
  readonly maxStaleRate: number;
  /** Minimum recovery success rate (0–1). */
  readonly minRecoverySuccessRate: number;
  /** Maximum crash rate (0–1) for mobile releases. */
  readonly maxCrashRate: number;
  /** p95 duration ceiling in ms for aisle finalize (0 = disabled). */
  readonly maxP95FinalizationMs: number;
  /** Queue depth warning / critical. */
  readonly queueDepthWarning: number;
  readonly queueDepthCritical: number;
  /** Oldest pending age seconds warning / critical. */
  readonly oldestPendingAgeWarningSec: number;
  readonly oldestPendingAgeCriticalSec: number;
}

export const DEFAULT_PRODUCTION_CUTOVER: ProductionCutoverThresholds = {
  minSuccessRate: 0.99,
  maxErrorRate: 0.01,
  maxDuplicateRate: 0,
  maxStaleRate: 0.02,
  minRecoverySuccessRate: 0.99,
  maxCrashRate: 0.005,
  maxP95FinalizationMs: 0,
  queueDepthWarning: 100,
  queueDepthCritical: 500,
  oldestPendingAgeWarningSec: 900,
  oldestPendingAgeCriticalSec: 3600,
};

function asFiniteNumber(value: unknown, fallback: number): number {
  if (typeof value === 'number' && Number.isFinite(value)) {
    return value;
  }
  if (typeof value === 'string' && value.trim()) {
    const n = Number(value);
    if (Number.isFinite(n)) {
      return n;
    }
  }
  return fallback;
}

function fromEnv(name: string): string {
  return typeof process !== 'undefined' && process.env?.[name] ? String(process.env[name]).trim() : '';
}

/** Resolve cutover thresholds from optional raw config + env overrides. */
export function resolveProductionCutover(raw: unknown): ProductionCutoverThresholds {
  const source = raw && typeof raw === 'object' ? (raw as Record<string, unknown>) : {};
  const num = (key: keyof ProductionCutoverThresholds, envKey: string, fallback: number): number => {
    const envVal = fromEnv(envKey);
    if (envVal) {
      return asFiniteNumber(envVal, fallback);
    }
    return asFiniteNumber(source[key], fallback);
  };
  return {
    minSuccessRate: num('minSuccessRate', 'DINAMIC_CUTOVER_MIN_SUCCESS_RATE', DEFAULT_PRODUCTION_CUTOVER.minSuccessRate),
    maxErrorRate: num('maxErrorRate', 'DINAMIC_CUTOVER_MAX_ERROR_RATE', DEFAULT_PRODUCTION_CUTOVER.maxErrorRate),
    maxDuplicateRate: num(
      'maxDuplicateRate',
      'DINAMIC_CUTOVER_MAX_DUPLICATE_RATE',
      DEFAULT_PRODUCTION_CUTOVER.maxDuplicateRate,
    ),
    maxStaleRate: num('maxStaleRate', 'DINAMIC_CUTOVER_MAX_STALE_RATE', DEFAULT_PRODUCTION_CUTOVER.maxStaleRate),
    minRecoverySuccessRate: num(
      'minRecoverySuccessRate',
      'DINAMIC_CUTOVER_MIN_RECOVERY_SUCCESS_RATE',
      DEFAULT_PRODUCTION_CUTOVER.minRecoverySuccessRate,
    ),
    maxCrashRate: num('maxCrashRate', 'DINAMIC_CUTOVER_MAX_CRASH_RATE', DEFAULT_PRODUCTION_CUTOVER.maxCrashRate),
    maxP95FinalizationMs: num(
      'maxP95FinalizationMs',
      'DINAMIC_CUTOVER_MAX_P95_FINALIZATION_MS',
      DEFAULT_PRODUCTION_CUTOVER.maxP95FinalizationMs,
    ),
    queueDepthWarning: num(
      'queueDepthWarning',
      'DINAMIC_ALERT_QUEUE_DEPTH_WARNING',
      DEFAULT_PRODUCTION_CUTOVER.queueDepthWarning,
    ),
    queueDepthCritical: num(
      'queueDepthCritical',
      'DINAMIC_ALERT_QUEUE_DEPTH_CRITICAL',
      DEFAULT_PRODUCTION_CUTOVER.queueDepthCritical,
    ),
    oldestPendingAgeWarningSec: num(
      'oldestPendingAgeWarningSec',
      'DINAMIC_ALERT_OLDEST_PENDING_WARNING_SEC',
      DEFAULT_PRODUCTION_CUTOVER.oldestPendingAgeWarningSec,
    ),
    oldestPendingAgeCriticalSec: num(
      'oldestPendingAgeCriticalSec',
      'DINAMIC_ALERT_OLDEST_PENDING_CRITICAL_SEC',
      DEFAULT_PRODUCTION_CUTOVER.oldestPendingAgeCriticalSec,
    ),
  };
}

export interface RolloutPauseSignals {
  readonly successRate: number;
  readonly errorRate: number;
  readonly duplicateRate: number;
  readonly staleRate: number;
  readonly recoverySuccessRate: number;
  readonly crashRate: number;
  readonly p95FinalizationMs?: number;
  readonly queueDepth?: number;
  readonly oldestPendingAgeSec?: number;
}

export type RolloutDecision = 'continue' | 'pause_warning' | 'pause_critical';

/** Evaluate whether rollout should auto-pause given live signals vs thresholds. */
export function evaluateRolloutPause(
  thresholds: ProductionCutoverThresholds,
  signals: RolloutPauseSignals,
): { readonly decision: RolloutDecision; readonly reasons: readonly string[] } {
  const reasons: string[] = [];
  let decision: RolloutDecision = 'continue';

  const bump = (level: RolloutDecision, reason: string) => {
    reasons.push(reason);
    if (level === 'pause_critical') {
      decision = 'pause_critical';
      return;
    }
    if (decision === 'continue') {
      decision = level;
    }
  };

  if (signals.successRate < thresholds.minSuccessRate) {
    bump('pause_critical', `successRate ${signals.successRate} < ${thresholds.minSuccessRate}`);
  }
  if (signals.errorRate > thresholds.maxErrorRate) {
    bump('pause_critical', `errorRate ${signals.errorRate} > ${thresholds.maxErrorRate}`);
  }
  if (signals.duplicateRate > thresholds.maxDuplicateRate) {
    bump('pause_critical', `duplicateRate ${signals.duplicateRate} > ${thresholds.maxDuplicateRate}`);
  }
  if (signals.staleRate > thresholds.maxStaleRate) {
    bump('pause_warning', `staleRate ${signals.staleRate} > ${thresholds.maxStaleRate}`);
  }
  if (signals.recoverySuccessRate < thresholds.minRecoverySuccessRate) {
    bump('pause_critical', `recoverySuccessRate ${signals.recoverySuccessRate} < ${thresholds.minRecoverySuccessRate}`);
  }
  if (signals.crashRate > thresholds.maxCrashRate) {
    bump('pause_critical', `crashRate ${signals.crashRate} > ${thresholds.maxCrashRate}`);
  }
  if (
    thresholds.maxP95FinalizationMs > 0 &&
    typeof signals.p95FinalizationMs === 'number' &&
    signals.p95FinalizationMs > thresholds.maxP95FinalizationMs
  ) {
    bump('pause_warning', `p95FinalizationMs ${signals.p95FinalizationMs} > ${thresholds.maxP95FinalizationMs}`);
  }
  if (typeof signals.queueDepth === 'number') {
    if (signals.queueDepth >= thresholds.queueDepthCritical) {
      bump('pause_critical', `queueDepth ${signals.queueDepth} >= ${thresholds.queueDepthCritical}`);
    } else if (signals.queueDepth >= thresholds.queueDepthWarning) {
      bump('pause_warning', `queueDepth ${signals.queueDepth} >= ${thresholds.queueDepthWarning}`);
    }
  }
  if (typeof signals.oldestPendingAgeSec === 'number') {
    if (signals.oldestPendingAgeSec >= thresholds.oldestPendingAgeCriticalSec) {
      bump(
        'pause_critical',
        `oldestPendingAgeSec ${signals.oldestPendingAgeSec} >= ${thresholds.oldestPendingAgeCriticalSec}`,
      );
    } else if (signals.oldestPendingAgeSec >= thresholds.oldestPendingAgeWarningSec) {
      bump(
        'pause_warning',
        `oldestPendingAgeSec ${signals.oldestPendingAgeSec} >= ${thresholds.oldestPendingAgeWarningSec}`,
      );
    }
  }

  return { decision, reasons };
}
